File: Interview-Pratice-FANNG/2-D-Array/dfs.js
/**
 * -> row
 * 1  |  2
 * 1  |  3
 * |
 * |
 * column 
 */

const data = require('./data');
const directions = require('./directions')


function dfs(matrix,explore,row,column,values) {
    
    // Check if two max 
    if(row < 0   || column < 0  || row >= matrix.length || column >= matrix[0].length || explore[row][column]) return;
    values.push(matrix[row][column])
    explore[row][column] = true;
    for(let i = 0 ; i < directions.length ; i++) {
        
        dfs(matrix, explore, row + directions[i][0], column + directions[i][1], values)
    }
}
const traversalDFs  = (matrix) => {
    const explore = new Array(matrix.length).fill(0).map(el => new Array(matrix[0].length).fill(false))
    const values = []
    dfs(matrix,explore, 0, 0, values)
    return values
}
console.log(traversalDFs(data))

File: Interview-Pratice-FANNG/2-D-Array/dfs.py
from data import data
from directions import directions
def dfs(data,values,row,column,seen):
  
    if row < 0 or column < 0 or row >= len(data) or column >= len(data[0]) or seen[row][column]:
        return

    val = data[row][column]
    values.append(val)
 
    seen[row][column] = True
    for direction in directions:
        next_row = row + direction[0]
        next_col = column + direction[1]
        dfs(data,values,next_row,next_col,seen)
def travel_dfs(data):
    values = []
    seen = [[False for i in range(len(data[0]))] for j in range(len(data))]
    dfs(data,values,0,0,seen)
    return values
  
print(travel_dfs(data))
